Firestore/core/src/firebase/firestore/model/field_value.cc

```diff
@@ -159,17 +367,55 @@ void FieldValue::SwitchTo(const Type type) {
   // Not same type. Destruct old type first and then initialize new type.
   // Must call destructor explicitly for any non-POD type.
   switch (tag_) {
+    case Type::Timestamp:
+      timestamp_value_.~Timestamp();
+      break;
+    case Type::ServerTimestamp:
+      server_timestamp_value_.~ServerTimestamp();
+      break;
+    case Type::String:
+      string_value_.~basic_string();
+      break;
+    case Type::Blob:
+      blob_value_.~Blob();
+      break;
+    case Type::GeoPoint:
+      geo_point_value_.~GeoPoint();
+      break;
     case Type::Array:
       array_value_.~vector();
       break;
+    case Type::Object:
+      object_value_.~map();
+      break;
     default: {}  // The other types where there is nothing to worry about.
   }
   tag_ = type;
   // Must call constructor explicitly for any non-POD type to initialize.
   switch (tag_) {
+    case Type::Timestamp:
+      new (&timestamp_value_) Timestamp(0, 0);
+      break;
+    case Type::ServerTimestamp:
+      // We initialize them to origin to avoid expensive calls to get `now'.
+      new (&server_timestamp_value_)
+          ServerTimestamp{Timestamp::Origin(), Timestamp::Origin()};
+      break;
+    case Type::String:
+      new (&string_value_) std::string();
+      break;
+    case Type::Blob:
+      new (&blob_value_) Blob(Blob::CopyFrom(nullptr, 0));
+      break;
+    case Type::GeoPoint:
+      new (&geo_point_value_) GeoPoint(0, 0);
```

Comment:
Types that aren't default-constructable aren't very user friendly if they're intended to be used as value types.

We should just make these have a default constructor so that we're not littering the callers with this kind of stuff.

Reply:
None of the other client SDK's has constructor of GeoPoint without parameters. On the other hand, the C++ language does make it different from other SDK, where variables are mostly reference. Let me add default constructor to GeoPoint for now; probably worth to discuss since this goes into the public api geo_point.h.